src/Infrastructure/LeanCode.Firebase.FCM/NotificationDataConverter.cs

```diff
@@ -0,0 +1,58 @@
+using System.Collections.Immutable;
+using System.Reflection;
+
+namespace LeanCode.Firebase.FCM;
+
+public sealed class NotificationDataConverter
+{
+    private const string TypeField = "Type";
+
+    private ImmutableDictionary<Type, Func<object, string>> formatters = ImmutableDictionary<
+        Type,
+        Func<object, string>
+    >.Empty;
+
+    public void FormatUsing<T>(Func<T, string> formatter)
+    {
+        formatters = formatters.Add(typeof(T), o => formatter((T)o));
+    }
+
+    /// <summary>
+    /// Converts POCO object to a notification data dictionary. Does not support hierarchical
+    /// data.
+    /// </summary>
+    public Dictionary<string, string> ToNotificationData(object data)
+    {
+        var type = data.GetType();
+
+        var result = new Dictionary<string, string>() { [TypeField] = type.Name, };
+
+        foreach (var prop in type.GetProperties(BindingFlags.Public | BindingFlags.Instance))
+        {
+            if (prop.Name != TypeField)
+            {
+                var value = prop.GetValue(data);
+
+                if (value is not null)
+                {
+                    var formatter = formatters.GetValueOrDefault(value.GetType(), DefaultValueFormatter);
```

Comment:
I would not go that way - let's make this class simple. We should re-model how we send pushes  either way (and it seems it will be sooner than later because of other requirements).

Reply:
Good catch, will see

Reply:
Remodeling would be nice indeed, but while we're still on this converter requiring users of this class to always register two formatters for nullable value types is error-prone. And fixing this is just one

```csharp
if (type.IsConstructedGenericType && type.GetGenericTypeDefinition() == typeof(Nullable<>))
{
    type = type.GenericTypeArguments.Single();
}
```

away.

Reply:
Yeah, I think you're right. We can go with this I think.